### PR TITLE
Block indent tuple type when fn_call_style is Block

### LINKED ISSUE #1662: Tuples should be block indented
With RFC style. Both literals and types.

### PR DESCRIPTION
**Pull request by topecongiro**
This PR closes #1662.

In order to reuse the code for rewriting tupe literals and tupe types, we need a way to distinguish `ast::Expr` and `ast::Ty`. This PR adds `ExprTypeIdentifier` trait to solve this, although I am not convinced that this is an ideal way, especially it requires reconstructing a vector of args.
```rust
         let args = args.iter().map(|e| e.to_expr()).collect::<Vec<_>>(); 
```

### CODE REVIEW

**Comment by topecongiro** [comment]
Rebased to resolve the conflicts with the current master.

File: src/expr.rs [hunk 1]
```diff
@@ -1921,7 +1926,22 @@ fn rewrite_call_args(
     // Try letting the last argument overflow to the next line with block
     // indentation. If its first line fits on one line with the other arguments,
     // we format the function arguments horizontally.
-    let tactic = try_overflow_last_arg(&item_context, &mut item_vec, args, shape, one_line_width);
+    let tactic = if args.len() > 0 && args[0].is_expr() {
+        let args = args.iter().map(|e| e.to_expr()).collect::<Vec<_>>();
```
Comment (nrc) [inline src/expr.rs]: Rather than use `is_expr` and an `if`, could you pass a closure from the client?

File: src/expr.rs [hunk 2]
```diff
@@ -2546,3 +2566,29 @@ fn rewrite_expr_addrof(
     };
     rewrite_unary_prefix(context, operator_str, expr, shape)
 }
+
+pub trait ExprTypeIdentifier {
+    fn is_type(&self) -> bool;
+    fn is_expr(&self) -> bool {
+        !self.is_type()
+    }
```
Comment (nrc) [inline src/expr.rs]: Why not implement is_expr directly and skip is_type? But see my previous comment for perhaps some way to eliminate is_expr altogether

File: src/expr.rs (hunk 1 above)
Comment (topecongiro) [inline src/expr.rs]: I think it is possible and have tried. Unfortunately the first attempt failed with `error[E0275]: overflow evaluating the requirement`.

**Comment by topecongiro** [comment]
I was unable to use closure, so instead used `filter_map` to avoid methods returning bool.

**Comment by nrc** [comment]
Thanks, looks improved!

**Pull request merged**